File: AlgorithmRise/extra/LeetCode/程序员面试宝典/面试题 04.02. 最小高度树/draft-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {

    public void buildTree(TreeNode node, int[] nums, int index){

        node.val = nums[index];
        int length = nums.length;

        // 数组中最后一个
        if(length-1 == index){
            return;
        }

        if(length-2 == index){
            node.left = new TreeNode(0);
            buildTree(node.left, nums, index+1);
        }else{
            node.left = new TreeNode( nums[index+1] );
            node.right = new TreeNode( nums[index+2] );
            buildTree(node.left, nums, index+1);
            buildTree(node.right, nums, index+2);
        }
    }

    public TreeNode sortedArrayToBST(int[] nums) {

        if(nums.length < 1){
            return null;
        }
        TreeNode node = new TreeNode(0);
        buildTree( node, nums, 0);

        return node;
    }
}

File: AlgorithmRise/extra/LeetCode/Algorithms/94. Binary Tree Inorder Traversal/draft-1.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function(root) {

    let stack = [root], result = [];

    while(stack.length){
                
        let length = stack.length;
        if(stack[length-1].left){
            stack.push(stack[length-1].left);
        }

        result.push(stack[length-1].val);

        if(stack[length-1].right){
            stack.push(stack[length-1].right);
        }

        stack.pop();
        console.log(stack.length);
    }

    return result;
};